### PR TITLE
[batch] Write user's JVM logs to disk

### PR DESCRIPTION
**Pull request by jigold**
Closes hail-is/hail-tasks#2

@danking Can you take a look at this before I start testing? I think the query/log4j.properties file still needs to be there to configure the logs that show up for the JVM and not the user's jobs.

### CODE REVIEW

**Comment by jigold** [comment]
Actually, it looks like you got rid of logs entirely for the JVM entryway?

File: hail/src/main/scala/is/hail/backend/service/Main.scala [hunk 1]
```diff
@@ -1,11 +1,34 @@
...
+    LogManager.resetConfiguration()
+    PropertyConfigurator.configure(logProps)
+  }
+
   def main(argv: Array[String]): Unit = {
-    argv(1) match {
+    val logFile = argv(1)
+    configureLogging(logFile)
+
+    argv(2) match {
```
Comment (danking) [inline hail/src/main/scala/is/hail/backend/service/Main.scala:31]: looks right

**Comment by jigold** [comment]
I deleted the `query` subdirectory. Ready for another check.

File: query/Makefile [hunk 2]
```diff
@@ -1,61 +0,0 @@
-include ../config.mk
```
Comment (danking) [inline query/Makefile:1]: We still need the makefile, it has targets for uploading a jar and running ipython or pytests with the right env vars. We might want a different solution long-term, but for now this is how that happens.

**Review by danking — APPROVED** [review]
Wow, this is great. I didn't realize it would be so succinct! Awesome.

**Review by danking — DISMISSED** [review]
oops meant to request changers

**Comment by jigold** [comment]
Great! I'll still test the changes by hand later if for no other reason to make sure the UI still works and I know how to use the query service.

**Comment by jigold** [comment]
Screen Shot 2022-03-03 at 12 06 23 PM

**Comment by danking** [comment]
ahhhh, this is so cool; I love it

**Review by danking — APPROVED** [review]

**Comment by danking** [comment]
I search the google logs for the pull request in question. For example,

```
(resource.labels.namespace_name="pr-11471-default-fn1xhr9ahy9v" AND
 labels."k8s-pod/app":"batch"
) OR (
labels."compute.googleapis.com/resource_name":"batch-worker-pr-11471-default-fn1xhr9ahy9v" AND
 logName:"worker.log")
```

That PR seems to have a worker stuck alive: `batch-worker-pr-11471-default-fn1xhr9ahy9v-standard-ux0sp`.

**Comment by danking** [comment]
From that VM, I can get into the container, install the jdk, then run jstack on one of the hung JVMs.
```
# jstack 1433
...
"pool-1-thread-1" #18 prio=5 os_prio=0 tid=0x00007f50c4f23000 nid=0x82c waiting on condition [0x00007f5084eeb000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000e8ddaea0> (a scala.concurrent.impl.Promise$CompletionLatch)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at scala.concurrent.impl.Promise$DefaultPromise.tryAwait(Promise.scala:242)
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:258)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:263)
	at scala.concurrent.Await$.$anonfun$result$1(package.scala:220)
	at scala.concurrent.Await$$$Lambda$2201/1092639564.apply(Unknown Source)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:57)
	at scala.concurrent.Await$.result(package.scala:146)
	at is.hail.backend.service.ServiceBackend.parallelizeAndComputeWithIndex(ServiceBackend.scala:145)
...
```

This is the line that waits to upload the compiled code for the workers to Google Cloud Storage.

The other threads appear to be waiting on the memory service:
```
"pool-2-thread-2" #27 prio=5 os_prio=0 tid=0x00007f5028ad9000 nid=0x88d waiting on condition [0x00007f50274fc000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at is.hail.services.package$.sleepAndBackoff(package.scala:32)
	at is.hail.services.package$.retryTransientErrors(package.scala:86)
	at is.hail.services.Requester.requestWithHandler(Requester.scala:69)
	at is.hail.services.Requester.request(Requester.scala:94)
	at is.hail.services.memory_client.MemoryClient.write(MemoryClient.scala:45)
	at is.hail.io.fs.ServiceCacheableFS$$anon$1.close(ServiceCacheableFS.scala:46)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
	at is.hail.utils.package$.using(package.scala:658)
	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$5(ServiceBackend.scala:127)
	at is.hail.backend.service.ServiceBackend$$Lambda$2194/482268176.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at is.hail.services.package$.retryTransientErrors(package.scala:77)
	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$4(ServiceBackend.scala:127)
	at is.hail.backend.service.ServiceBackend$$Lambda$2191/716305671.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.concurrent.Future$$$Lambda$2188/1126720330.apply(Unknown Source)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.Future$$Lambda$2189/609808342.apply(Unknown Source)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.Promise$$Lambda$2190/183883584.apply(Unknown Source)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

"pool-2-thread-1" #26 prio=5 os_prio=0 tid=0x00007f502866e000 nid=0x88c waiting on condition [0x00007f50275fd000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at is.hail.services.package$.sleepAndBackoff(package.scala:32)
	at is.hail.services.package$.retryTransientErrors(package.scala:86)
	at is.hail.services.Requester.requestWithHandler(Requester.scala:69)
	at is.hail.services.Requester.request(Requester.scala:94)
	at is.hail.services.memory_client.MemoryClient.write(MemoryClient.scala:45)
	at is.hail.io.fs.ServiceCacheableFS$$anon$1.close(ServiceCacheableFS.scala:46)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
	at java.io.ObjectOutputStream$BlockDataOutputStream.close(ObjectOutputStream.java:1828)
	at java.io.ObjectOutputStream.close(ObjectOutputStream.java:742)
	at is.hail.utils.package$.using(package.scala:658)
	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$2(ServiceBackend.scala:119)
	at is.hail.backend.service.ServiceBackend$$Lambda$2192/1687163734.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at is.hail.services.package$.retryTransientErrors(package.scala:77)
	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$1(ServiceBackend.scala:119)
	at is.hail.backend.service.ServiceBackend$$Lambda$2187/1214727901.apply$mcV$sp(Unknown Source)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
	at scala.concurrent.Future$$$Lambda$2188/1126720330.apply(Unknown Source)
	at scala.util.Success.$anonfun$map$1(Try.scala:255)
	at scala.util.Success.map(Try.scala:213)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
	at scala.concurrent.Future$$Lambda$2189/609808342.apply(Unknown Source)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
	at scala.concurrent.impl.Promise$$Lambda$2190/183883584.apply(Unknown Source)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by danking** [comment]
Memory got stuck offline due to the pod being offline: https://cloudlogging.app.goo.gl/mQKPXPCxNcsVEw5eA

**Comment by danking** [comment]
The memory image is cleaned up too eagerly. https://ci.hail.is/batches/733555/jobs/181

**Comment by danking** [comment]
Ah, service backend tests need to depend on deploy_memory

**Comment by jigold** [comment]
I made a PR for that already

**Pull request merged**